x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/CoordinatorProxyAction.java

```diff
@@ -0,0 +1,140 @@
+            coordinator.schedule(request, listener);
+        }
+    }
+
+    public static class Coordinator {
+
+        final Client client;
+        final int maxLookupsPerRequest;
+        final int maxNumberOfConcurrentRequests;
+        final BlockingQueue<Item> queue = new LinkedBlockingQueue<>();
+
+        int numberOfOutstandingRequests = 0;
+
+        public Coordinator(Client client, Settings settings) {
+            this(client,
+                EnrichPlugin.COORDINATOR_PROXY_MAX_LOOKUPS_PER_REQUEST.get(settings),
+                EnrichPlugin.COORDINATOR_PROXY_MAX_CONCURRENT_REQUESTS.get(settings));
+        }
+
+        Coordinator(Client client, int maxLookupsPerRequest, int maxNumberOfConcurrentRequests) {
+            this.client = client;
+            this.maxLookupsPerRequest = maxLookupsPerRequest;
+            this.maxNumberOfConcurrentRequests = maxNumberOfConcurrentRequests;
+        }
+
+        void schedule(SearchRequest searchRequest, ActionListener<SearchResponse> listener) {
+            queue.add(new Item(searchRequest, listener));
+            coordinateLookups();
+        }
+
+        synchronized void coordinateLookups() {
+            while (queue.isEmpty() == false &&
+                numberOfOutstandingRequests <= maxNumberOfConcurrentRequests) {
+
+                final List<Item> items = new LinkedList<>();
+                queue.drainTo(items, maxLookupsPerRequest);
+                final MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+                items.forEach(item -> multiSearchRequest.add(item.searchRequest));
+
+                numberOfOutstandingRequests++;
+                client.multiSearch(multiSearchRequest, ActionListener.wrap(response -> {
+                    handleResponse(items, response, null);
+                }, e -> {
+                    handleResponse(items, null, e);
+                }));
+            }
+        }
+
+        synchronized void handleResponse(List<Item> items, MultiSearchResponse response, Exception e) {
+            numberOfOutstandingRequests--;
+
+            if (response != null) {
+                assert items.size() == response.getResponses().length;
+                for (int i = 0; i < response.getResponses().length; i++) {
+                    MultiSearchResponse.Item responseItem = response.getResponses()[i];
+                    Item item = items.get(i);
+
+                    if (responseItem.isFailure()) {
+                        item.actionListener.onFailure(responseItem.getFailure());
+                    } else {
+                        item.actionListener.onResponse(responseItem.getResponse());
+                    }
+                }
+            } else if (e != null) {
+                items.forEach(item -> item.actionListener.onFailure(e));
+            } else {
+                assert false;
+            }
+
+            coordinateLookups();
```

Comment:
I'm assuming we need to call it here because we just decreased the number of outstanding requests, so this might have made room for new requests? If so can you leave a comment to make it clear?